File: wow-minmax/src/main/java/wow/minmax/model/CharacterId.java
package wow.minmax.model;

import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import wow.commons.model.character.CreatureType;
import wow.commons.model.pve.PhaseId;

import java.io.Serializable;
import java.util.Objects;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * User: POlszewski
 * Date: 2023-04-01
 */
@AllArgsConstructor
@Getter
@EqualsAndHashCode
public class CharacterId implements Serializable {
	private final UUID profileId;
	private final PhaseId phaseId;
	private final int level;
	private final CreatureType enemyType;
	private final int enemyLevelDiff;

	private static final String SEPARATOR = ",";

	public static CharacterId parse(String value) {
		if (value == null) {
			return null;
		}
		String[] parts = value.split(SEPARATOR);
		return new CharacterId(
				UUID.fromString(parts[0]),
				PhaseId.parse(parts[1]),
				Integer.parseInt(parts[2]),
				CreatureType.parse(parts[3]),
				Integer.parseInt(parts[4])
		);
	}

	@Override
	public String toString() {
		return Stream.of(profileId, phaseId, level, enemyType, enemyLevelDiff)
				.map(Objects::toString)
				.collect(Collectors.joining(SEPARATOR))
				.toLowerCase();
	}
}

File: apps/wow-minmax-ui/src/app/modules/character/model/CharacterId.ts
import { PhaseId } from "../../shared/model/character/PhaseId";

export interface CharacterIdParts {
	profileId: string;
	phaseId: PhaseId;
	level: number
	enemyTypeId: string;
	enemyLevelDiff: number;
}

const SEPARATOR = ',';

export function formatCharacterId(idParts: CharacterIdParts) {
	return (idParts.profileId + SEPARATOR +
			idParts.phaseId + SEPARATOR +
			idParts.level + SEPARATOR +
			idParts.enemyTypeId + SEPARATOR +
			idParts.enemyLevelDiff).toLowerCase();
}

export function parseCharacterId(value: string): CharacterIdParts {
	const parts = value.split(SEPARATOR);

	return {
			profileId: parts[0],
			phaseId: parts[1] as PhaseId,
			level: Number.parseInt(parts[2]),
			enemyTypeId: parts[3],
			enemyLevelDiff: Number.parseInt(parts[4])
	};
}
